src/app/clusters/operational-state-server/operational-state-delegate.h

```diff
@@ -0,0 +1,323 @@
+
+        if (details.HasValue())
+        {
+            memset(mErrorStateDetailsBuffer, 0, sizeof(mErrorStateDetailsBuffer));
+            if (details.Value().size() > sizeof(mErrorStateDetailsBuffer))
+            {
+                memcpy(mErrorStateDetailsBuffer, details.Value().data(), sizeof(mErrorStateDetailsBuffer));
+                errorStateDetails.SetValue(CharSpan(mErrorStateDetailsBuffer, sizeof(mErrorStateDetailsBuffer)));
+            }
+            else
+            {
+                memcpy(mErrorStateDetailsBuffer, details.Value().data(), details.Value().size());
+                errorStateDetails.SetValue(CharSpan(mErrorStateDetailsBuffer, details.Value().size()));
+            }
+        }
+        else
+        {
+            errorStateDetails = NullOptional;
+        }
+    }
+
+private:
+    char mErrorStateLabelBuffer[kOperationalErrorLabelMaxSize];
+    char mErrorStateDetailsBuffer[kOperationalErrorDetailsMaxSize];
+};
+
+/**
+ * A class which represents the operational phase of an Operational State cluster derivation instance.
+ */
+struct GenericOperationalPhase
+{
+    GenericOperationalPhase(app::DataModel::Nullable<CharSpan> name) { Set(name); }
+
+    GenericOperationalPhase(const GenericOperationalPhase & ph) { *this = ph; }
+
+    GenericOperationalPhase & operator=(const GenericOperationalPhase & ph)
+    {
+        Set(ph.mPhaseName);
+        return *this;
+    }
+
+    bool IsMissing() const { return mPhaseName.IsNull(); }
+    app::DataModel::Nullable<CharSpan> mPhaseName;
+
+private:
+    void Set(app::DataModel::Nullable<CharSpan> name)
+    {
+        if (name.IsNull())
+        {
+            mPhaseName.SetNull();
+        }
+        else
+        {
+            memset(mPhaseNameBuffer, 0, sizeof(mPhaseNameBuffer));
+            if (name.Value().size() > sizeof(mPhaseNameBuffer))
+            {
+                memcpy(mPhaseNameBuffer, name.Value().data(), sizeof(mPhaseNameBuffer));
+                mPhaseName = app::DataModel::Nullable<CharSpan>(CharSpan(mPhaseNameBuffer, sizeof(mPhaseNameBuffer)));
+            }
+            else
+            {
+                memcpy(mPhaseNameBuffer, name.Value().data(), name.Value().size());
+                mPhaseName = app::DataModel::Nullable<CharSpan>(CharSpan(mPhaseNameBuffer, name.Value().size()));
+            }
+        }
+    }
+
+    char mPhaseNameBuffer[kOperationalPhaseNameMaxSize];
+};
+
+/**
+ * A class which represents the operational completion of an Operational State cluster derivation instance.
+ */
+struct GenericOperationCompletion : public app::Clusters::OperationalState::Events::OperationCompletion::Type
+{
+    GenericOperationCompletion(
+        uint8_t aCompletionErrorCode,
+        const Optional<DataModel::Nullable<uint32_t>> & aTotalOperationalTime = Optional<DataModel::Nullable<uint32_t>>::Missing(),
+        const Optional<DataModel::Nullable<uint32_t>> & aPausedTime           = Optional<DataModel::Nullable<uint32_t>>::Missing())
```

Comment:
```suggestion
        const Optional<DataModel::Nullable<uint32_t>> & aTotalOperationalTime = NullOptional,
        const Optional<DataModel::Nullable<uint32_t>> & aPausedTime           = NullOptional)
```
may be more readable, but either way.